xds/internal/resolver/serviceconfig.go

```diff
@@ -314,16 +314,13 @@ type interceptorList struct {
 	interceptors []iresolver.ClientInterceptor
 }
 
-func (il *interceptorList) NewStream(ctx context.Context, ri iresolver.RPCInfo, cs iresolver.ClientStream) (context.Context, iresolver.ClientStream, error) {
-	for _, i := range il.interceptors {
-		var err error
-		newCTX, newCS, err := i.NewStream(ctx, ri, cs)
-		if err != nil {
-			cs.Done()
-			return nil, nil, err
+func (il *interceptorList) NewStream(ctx context.Context, ri iresolver.RPCInfo, done func(), newStream func(ctx context.Context, done func()) (iresolver.ClientStream, error)) (iresolver.ClientStream, error) {
```

Comment:
> As discussed, there's a way to do this without done.

There is... now the question is whether it's worth doing.  On the one hand, it's a smaller API.  On the other hand, it's harder to use.

> And also, add a unit test for interceptorList?

The order is confirmed by the tests I added (`TestXDSResolverHTTPFilters`), which has multiple filters.  I can add a more comprehensive / unit-style test if you see things it's not covering at that level.

Reply:
This API is currently internal. Will we ever export it?
I think if it's exported, we want to keep the API smaller.

And if we export it later, how to we remember change this? Would a comment be enough?

Reply:
`TestXDSResolverHTTPFilters` is nondeterministic due to the map.
I was thinking a unit test by creating a `interceptorList`, and call it's `NewStream`. So we know the interceptors are executed in the expected order.

Reply:
> This API is currently internal. Will we ever export it?

It seems unlikely, but never say never.

> I think if it's exported, we want to keep the API smaller.

I'm not sure in this case.  Ease of use is more important than size.

> And if we export it later, how to we remember change this? Would a comment be enough?

If we decide to make an API public, it needs to be fully evaluated at that time.  And go through an experimental stage.  I'm not concerned about forgetting something here.

> TestXDSResolverHTTPFilters is nondeterministic due to the map.

I don't think so.  Filter order is deterministic.  Unless I'm misunderstanding.

> I was thinking a unit test by creating a interceptorList, and call it's NewStream. So we know the interceptors are executed in the expected order.

That's what's happening in that test, but it's going through additional logic as well.

Reply:
>> I think if it's exported, we want to keep the API smaller.
>
> I'm not sure in this case. Ease of use is more important than size.

The argument is, it's ease of use for `interceptorList` to chain the interceptors, not ease of use for the end users.
It's more confusing for the users (who implement an interceptor), because we give them a `done()`, but also say "don't call it, something goes wrong if you call it, just wrap it in your new `done`, give it back to the `newStream()`".

Should be OK now, since it's an internal API. We can argue about it more if we ever want to make it public.

Reply:
> We can argue about it more if we ever want to make it public.

Sounds good.

For future views, I disagree that it's confusing to the interceptor to give it a function that needs to be called when the stream is complete.  No wrapping is necessary unless the interceptor itself needs a function called when the stream is complete.  A normal interceptor is much simpler to implement than `interceptorList`, as it only has to worry about itself.  The `done` wrapping is no different from wrapping an interceptor/struct that has a `Close` method.

For an example of typical `NewStream` usage, see the fault injection filter.